homeassistant/components/samsungtv/media_player.py

```diff
@@ -239,33 +243,31 @@ async def _async_startup_app_list(self) -> None:
                 "Failed to load app list from %s: %s", self._host, err.__repr__()
             )
 
-    async def _async_startup_upnp(self) -> None:
+    async def _async_startup_dmr(self) -> None:
         assert self._ssdp_rendering_control_location is not None
-        if self._upnp_device is None:
+        if self._dmr_device is None:
             session = async_get_clientsession(self.hass)
             upnp_requester = AiohttpSessionRequester(session)
             upnp_factory = UpnpFactory(upnp_requester)
+            upnp_device: UpnpDevice | None = None
             with contextlib.suppress(UpnpConnectionError):
-                self._upnp_device = await upnp_factory.async_create_device(
+                upnp_device = await upnp_factory.async_create_device(
                     self._ssdp_rendering_control_location
                 )
-
-    def _get_upnp_service(self, log: bool = False) -> UpnpService | None:
-        if self._upnp_device is None:
-            if log:
-                LOGGER.info("Upnp services are not available on %s", self._host)
-            return None
-
-        if service := self._upnp_device.services.get(UPNP_SVC_RENDERING_CONTROL):
-            return service
-
-        if log:
-            LOGGER.info(
-                "Upnp service %s is not available on %s",
-                UPNP_SVC_RENDERING_CONTROL,
-                self._host,
+            if not upnp_device:
+                return
+            _, event_ip = await async_get_local_ip(
+                self._ssdp_rendering_control_location, self.hass.loop
             )
-        return None
+            source = (event_ip or "0.0.0.0", 0)
+            self._upnp_server = AiohttpNotifyServer(
+                requester=upnp_requester,
+                source=source,
+                callback_url=None,
```

Comment:
Can we get a callback when the Tv is muted or volume changed?

Reply:
If my understanding is correct, the upnp client will use an internal callback to set internal property which we then access.
I am not sure we need an extra callback, but maybe I am wrong in my understanding of the library.

Reply:
I *think* this pattern will allow getting events from the device and avoid the need to poll
https://github.com/home-assistant/core/blob/dev/homeassistant/components/dlna_dmr/media_player.py#L339